src/lavinmq/mqtt/client.cr

```diff
@@ -0,0 +1,227 @@
+      end
+
+      private def read_loop
+        loop do
+          @log.trace { "waiting for packet" }
+          packet = read_and_handle_packet
+          # The disconnect packet has been handled and the socket has been closed.
+          # If we dont breakt the loop here we'll get a IO/Error on next read.
+          break if packet.is_a?(MQTT::Disconnect)
+        end
+      rescue ex : ::MQTT::Protocol::Error::PacketDecode
+        @socket.close
+      rescue ex : MQTT::Error::Connect
+        @log.warn { "Connect error #{ex.inspect}" }
+      rescue ex : ::IO::Error
+        @log.warn(exception: ex) { "Read Loop error" }
+        publish_will if @will
+      rescue ex
+        publish_will if @will
+        raise ex
+      ensure
+        @broker.disconnect_client(client_id)
+        @socket.close
+        # move to disconnect client
+        @broker.vhost.rm_connection(self)
+      end
+
+      def read_and_handle_packet
+        packet : MQTT::Packet = MQTT::Packet.from_io(@io)
+        @log.trace { "Recieved packet:  #{packet.inspect}" }
+        @recv_oct_count += packet.bytesize
+
+        case packet
+        when MQTT::Publish     then recieve_publish(packet)
+        when MQTT::PubAck      then recieve_puback(packet)
+        when MQTT::Subscribe   then recieve_subscribe(packet)
+        when MQTT::Unsubscribe then recieve_unsubscribe(packet)
+        when MQTT::PingReq     then receive_pingreq(packet)
+        when MQTT::Disconnect  then return packet
+        else                        raise "received unexpected packet: #{packet}"
+        end
+        packet
+      end
+
+      def send(packet)
+        @lock.synchronize do
+          packet.to_io(@io)
+          @socket.flush
+        end
+        @send_oct_count += packet.bytesize
+      end
+
+      def receive_pingreq(packet : MQTT::PingReq)
+        send MQTT::PingResp.new
+      end
+
+      def recieve_publish(packet : MQTT::Publish)
+        @broker.publish(packet)
+        # Ok to not send anything if qos = 0 (at most once delivery)
+        if packet.qos > 0 && (packet_id = packet.packet_id)
+          send(MQTT::PubAck.new(packet_id))
+        end
+      end
+
+      def recieve_puback(packet)
+        @broker.sessions[@client_id].ack(packet)
+      end
+
+      def recieve_subscribe(packet : MQTT::Subscribe)
+        qos = @broker.subscribe(self, packet)
+        session = @broker.sessions[@client_id]
+        send(MQTT::SubAck.new(qos, packet.packet_id))
+      end
+
+      def recieve_unsubscribe(packet)
+        session = @broker.sessions[@client_id]
+        @broker.unsubscribe(self, packet)
+        send(MQTT::UnsubAck.new(packet.packet_id))
+      end
+
+      def details_tuple
+        {
+          vhost:     @broker.vhost.name,
+          user:      @user.name,
+          protocol:  "MQTT",
+          client_id: @client_id,
+        }.merge(stats_details)
+      end
+
+      private def publish_will
+        if will = @will
+          @broker.publish(will)
```

Comment:
I think we should create a `Publish` packet here and pass it to `Broker#publish` instead of having `Broker#publish` handling a union of `Publish | Will`.

Reply:
but then that id will be used and acked on the client, so might be dangerous

Reply:
ah nvm the id can be nil!